core/arch/arm/kernel/thread_spmc.c

```diff
@@ -326,46 +327,107 @@ static bool is_my_uuid(uint32_t w0, uint32_t w1, uint32_t w2, uint32_t w3)
+	fpi->execution_context = execution_context;
 
-	if (!is_nil_uuid(args->a1, args->a2, args->a3, args->a4) &&
-	    !is_my_uuid(args->a1, args->a2, args->a3, args->a4)) {
-		ret_fid = FFA_ERROR;
-		rc = FFA_INVALID_PARAMETERS;
-		goto out;
+	/*
+	 * BIT(0): Supports receipt of direct requests
+	 * BIT(1): Can send direct requests
+	 * BIT(2): Cannot send and receive indirect messages
+	 * BIT(3): Does not support receipt of notifications
+	 * BIT(4-5): Partition ID is a PE endpoint ID
+	 */
+	fpi->partition_properties = FFA_PARTITION_DIRECT_REQ_RECV_SUPPORT |
+				    FFA_PARTITION_DIRECT_REQ_RECV_SUPPORT;
+}
+
+static uint32_t handle_partition_info_get_all(size_t *elem_count,
+					      struct ffa_rxtx *rxtx)
+{
+	struct ffa_partition_info *fpi = rxtx->tx;
+
+	/* Add OP-TEE SP */
+	spmc_fill_partition_entry(fpi, my_endpoint_id, 0);
+	rxtx->tx_is_mine = false;
+	*elem_count = 1;
+	fpi++;
+
+	if (IS_ENABLED(CFG_SECURE_PARTITION)) {
+		size_t elements = (rxtx->size / sizeof(*fpi)) - 1;
+
+		if (sp_get_partitions_info(fpi, &elements))
+			return TEE_ERROR_SHORT_BUFFER;
+		elem_count += elements;
 	}
+	return FFA_SUCCESS_32;
+}
 
-	cpu_spin_lock(&rxtx->spinlock);
-	if (rxtx->size && rxtx->tx_is_mine) {
-		struct ffa_partition_info *fpi = rxtx->tx;
+void spmc_handle_partition_info_get(struct thread_smc_args *args,
+				    struct ffa_rxtx *rxtx)
+{
+	uint32_t ret_fid = FFA_ERROR;
+	uint32_t rc = 0;
+	uint32_t endpoint_id = my_endpoint_id;
+	struct ffa_partition_info *fpi = NULL;
 
-		fpi->id = my_endpoint_id;
-		fpi->execution_context = CFG_TEE_CORE_NB_CORE;
-		/*
-		 * Supports receipt of direct requests.
-		 * Can send direct requests.
-		 */
-		fpi->partition_properties = BIT(0) | BIT(1);
+	cpu_spin_lock(&rxtx->spinlock);
 
-		ret_fid = FFA_SUCCESS_32;
-		rc = 1;
-		rxtx->tx_is_mine = false;
-	} else {
-		ret_fid = FFA_ERROR;
+	if (!rxtx->size || !rxtx->tx_is_mine) {
 		if (rxtx->size)
 			rc = FFA_BUSY;
 		else
 			rc = FFA_DENIED; /* TX buffer not setup yet */
+		goto out;
+	}
+
+	fpi = rxtx->tx;
+
+	if (is_nil_uuid(args->a1, args->a2, args->a3, args->a4)) {
+		size_t elem_count = 0;
+
+		ret_fid = handle_partition_info_get_all(&elem_count, rxtx);
+		rc = elem_count;
+		goto out;
 	}
-	cpu_spin_unlock(&rxtx->spinlock);
+
+	if (!is_my_uuid(args->a1, args->a2, args->a3, args->a4)) {
+		if (IS_ENABLED(CFG_SECURE_PARTITION)) {
+			uint32_t uuid_array[4] = { 0 };
+			TEE_UUID uuid = { };
+			TEE_Result res = TEE_SUCCESS;
+
+			uuid_array[0] = args->a1;
+			uuid_array[1] = args->a2;
+			uuid_array[2] = args->a3;
+			uuid_array[3] = args->a4;
+			tee_uuid_from_octets(&uuid, (uint8_t *)uuid_array);
+
+			res = sp_find_session_id(&uuid, &endpoint_id);
+			if (res != TEE_SUCCESS) {
+				ret_fid = FFA_ERROR;
+				rc = FFA_INVALID_PARAMETERS;
+				goto out;
+			}
+		} else {
+			ret_fid = FFA_ERROR;
+			rc = FFA_INVALID_PARAMETERS;
+			goto out;
+		}
+	}
+
+	spmc_fill_partition_entry(fpi, endpoint_id, 1);
```

Comment:
This isn't correct in the case where `endpoint_id` is still `my_endpoint_id`

Reply:
Done